File: src/utils/nextstates.go
package utils

import (
	"slices"

	"github.com/utkarsh-naman/3T/src/Constants"
	"github.com/utkarsh-naman/3T/src/SymmetryReduction"
	"github.com/utkarsh-naman/3T/src/model"
)

func NextMoves(state model.State) []model.State {
	var NextSates []model.State
	continuum, tie, won, lost, turn, x, o, vacancy := model.Extract(state)

	if continuum != 1 {
		return NextSates
	}

	for i := 0; i < 9; i++ {
		var mask model.State = 1 << i
		if vacancy&mask != 0 {
			var nextstate model.State

			if turn == 0 { // X turn
				var board model.State = ((x ^ mask) << 18) | (o << 9) | (vacancy ^ mask)
				temporary := makeState(continuum, tie, won, lost, turn^1, board)
				nextstate = SymmetryReduction.CollapseEquivalents(finalTouch(temporary))
			} else {
				var board model.State = (x << 18) | ((o ^ mask) << 9) | (vacancy ^ mask)
				temporary := makeState(continuum, tie, won, lost, turn^1, board)
				nextstate = SymmetryReduction.CollapseEquivalents(finalTouch(temporary))
			}

			if !slices.Contains(NextSates, nextstate) {
				NextSates = append(NextSates, nextstate)
			}
		}
	}
	return NextSates
}

func wonby(player model.State) bool {
	for _, winmask := range Constants.WinMasks {
		if player&winmask == winmask {
			return true
		}
	}
	return false
}

func endAsTie(state model.State) bool {
	_, _, _, _, _, x, o, v := model.Extract(state)

	if v == 0 && !wonby(x) && !wonby(o) {
		return true
	}
	return false
}

func continueOrNot(state model.State) bool {
	_, _, _, _, _, x, o, v := model.Extract(state)

	if v == 0 || wonby(x) || wonby(o) {
		return false
	}
	return true
}

func finalTouch(temporary model.State) model.State {
	c, ti, w, l, tu, x, o, v := model.Extract(temporary)
	if continueOrNot(temporary) {
		c = 1
	} else {
		c = 0
	}

	if endAsTie(temporary) {
		ti = 1
	} else {
		ti = 0
	}

	if wonby(x) {
		w = 0
		l = 1
	}
	if wonby(o) {
		w = 1
		l = 0
	}

	return makeState(c, ti, w, l, tu, x<<18|o<<9|v)

}

func makeState(continuum, tie, won, lost, turn, board model.State) model.State {
	return continuum<<31 | tie<<30 | won<<29 | lost<<28 | turn<<27 | board
}

File: Rust/src/utils/next_states.rs
use crate::model::state_model::{extract, gen_board, State};
use crate::symmetry_reduction::symmred::collapse;

pub fn next_states(state: &State) ->Vec<State>{
    let mut next_moves:Vec<State> = Vec::new();
    let (continuum, _, _, _, turn, x, o, vacancy) = extract(&state);
    
    if continuum == 0{
        return next_moves;
    }

    let mut i0mask:u32;
    for i in 0..9{
        i0mask = 1<<i;
        if vacancy&i0mask != 0 {
            let next_state: State;

            if turn == 0 {
                let xoboard: State = ((x^i0mask)<<9) | o;
                next_state = collapse(&gen_board(&xoboard));
            } else {
                let xoboard: State = (x << 9) | (o^i0mask);
                next_state = collapse(&gen_board(&xoboard));
            }

            if !next_moves.contains(&next_state){
                next_moves.push(next_state);
            }
        }
    }
    next_moves
}